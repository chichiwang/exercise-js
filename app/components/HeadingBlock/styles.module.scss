@use "sass:map";
@use "app/var.module";

$padding-unit: map.get(var.$content, 'padding');

.headingBlock {
  margin: calc(2 * $padding-unit) 0;
  padding: 0 $padding-unit;
  width: 100%;
}
