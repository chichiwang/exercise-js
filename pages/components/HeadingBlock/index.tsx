import React from 'react';
import classNames from 'classnames';

import styles from './styles.module.scss';

type HeadingLevel = '1' | '2' | '3' | '4' | '5' | '6';

interface HeadingBlockProps extends React.PropsWithChildren {
  heading: string,
  level: HeadingLevel,
  center?: boolean,
}

function createHeading(heading: string, level: HeadingLevel, center: boolean) {
  return function DynamicHeading() {
    return React.createElement(
      `h${level}`,
      {
        className: classNames({
          [styles.center]: center,
        }),
      },
      heading,
    );
  };
}

function HeadingBlock({
  heading,
  level,
  center = false,
  children,
}: HeadingBlockProps) {
  const SectionHeading = createHeading(heading, level, center);

  return (
    <section className={styles.headingBlock}>
      <SectionHeading />
      <article>
        {children}
      </article>
    </section>
  );
}

export default HeadingBlock;
