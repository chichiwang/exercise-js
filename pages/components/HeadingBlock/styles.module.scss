@use "sass:map";
@use "app/var.module";

$padding-unit: map.get(var.$content, 'padding');

.headingBlock {
  margin: 0 auto;
  padding: calc(2 * $padding-unit) $padding-unit;
  max-width: map.get(var.$content, 'article-max-width');

  h1 {
    font-size: 2rem;
  }
}

.center {
  text-align: center;
}
