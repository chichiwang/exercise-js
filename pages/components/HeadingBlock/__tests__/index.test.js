import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import HeadingBlock from '../index';

describe('HeadingBlock Component', () => {
  test('matches uncentered snapshot', () => {
    const tree = renderer.create(
      <HeadingBlock heading="Test Heading" level="1">
        <p>Block content</p>
      </HeadingBlock>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('matches centered snapshot', () => {
    const tree = renderer.create(
      <HeadingBlock heading="Test Heading" level="1" center>
        <p>Block content</p>
      </HeadingBlock>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('renders child components', () => {
    const childText = 'Paragraph Contents';

    render(
      <HeadingBlock heading="Test Heading" level="1" center>
        <p>{childText}</p>
      </HeadingBlock>,
    );

    expect(screen.getByText(childText)).toBeInTheDocument();
  });

  describe('heading levels', () => {
    test('contains a h1 tag with the project name', () => {
      const headingText = 'H1 tag contents';

      render(
        <HeadingBlock heading={headingText} level="1" center>
          <p>Block content</p>
        </HeadingBlock>,
      );

      expect(screen.getByRole('heading', { level: 1 })).toHaveTextContent(headingText);
    });

    test('contains a h2 tag with the project name', () => {
      const headingText = 'H2 tag contents';

      render(
        <HeadingBlock heading={headingText} level="2" center>
          <p>Block content</p>
        </HeadingBlock>,
      );

      expect(screen.getByRole('heading', { level: 2 })).toHaveTextContent(headingText);
    });

    test('contains a h3 tag with the project name', () => {
      const headingText = 'H3 tag contents';

      render(
        <HeadingBlock heading={headingText} level="3" center>
          <p>Block content</p>
        </HeadingBlock>,
      );

      expect(screen.getByRole('heading', { level: 3 })).toHaveTextContent(headingText);
    });

    test('contains a h4 tag with the project name', () => {
      const headingText = 'H4 tag contents';

      render(
        <HeadingBlock heading={headingText} level="4" center>
          <p>Block content</p>
        </HeadingBlock>,
      );

      expect(screen.getByRole('heading', { level: 4 })).toHaveTextContent(headingText);
    });

    test('contains a h5 tag with the project name', () => {
      const headingText = 'H5 tag contents';

      render(
        <HeadingBlock heading={headingText} level="5" center>
          <p>Block content</p>
        </HeadingBlock>,
      );

      expect(screen.getByRole('heading', { level: 5 })).toHaveTextContent(headingText);
    });

    test('contains a h6 tag with the project name', () => {
      const headingText = 'H6 tag contents';

      render(
        <HeadingBlock heading={headingText} level="6" center>
          <p>Block content</p>
        </HeadingBlock>,
      );

      expect(screen.getByRole('heading', { level: 6 })).toHaveTextContent(headingText);
    });
  });
});
