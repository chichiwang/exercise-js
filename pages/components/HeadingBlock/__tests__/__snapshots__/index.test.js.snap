// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeadingBlock Component matches centered snapshot 1`] = `
<section
  className="headingBlock"
>
  <h1
    className="center"
  >
    Test Heading
  </h1>
  <article>
    <p>
      Block content
    </p>
  </article>
</section>
`;

exports[`HeadingBlock Component matches uncentered snapshot 1`] = `
<section
  className="headingBlock"
>
  <h1
    className=""
  >
    Test Heading
  </h1>
  <article>
    <p>
      Block content
    </p>
  </article>
</section>
`;
